In [14]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
# Flights from Houston,TX data set
flights = pd.read_csv("https://github.com/kiat/Elements-of-Data-Analytics/blob/main/datasets/airline/HoustonAirline.csv?raw=true") 
flights

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0       2008      1           4          5   1910.0        1910   2025.0   
1       2008      1           4          5   1345.0        1345   1453.0   
2       2008      1           4          5    736.0         735    839.0   
3       2008      1           4          5   1603.0        1550   1647.0   
4       2008      1           4          5   2105.0        2105   2151.0   
...      ...    ...         ...        ...      ...         ...      ...   
241100  2008     12          12          5   1122.0        1040   1418.0   
241101  2008     12          12          5    546.0         545   1011.0   
241102  2008     12          12          5   1114.0        1050   1406.0   
241103  2008     12          13          6   1612.0        1614   1910.0   
241104  2008     12          13          6   1222.0        1050   1519.0   

        CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0             2025            WN        823  ...    6.0     11.0          0   
1             1500            WN        901  ...    3.0     14.0          0   
2              850            WN        957  ...    3.0      9.0          0   
3             1635            WN        231  ...    5.0     10.0          0   
4             2150            WN        570  ...    7.0     11.0          0   
...            ...           ...        ...  ...    ...      ...        ...   
241100        1342            DL         50  ...    9.0     15.0          0   
241101         849            DL        668  ...   16.0     86.0          0   
241102        1349            DL       1292  ...   11.0      8.0          0   
241103        1919            DL          8  ...   16.0     18.0          0   
241104        1346            DL         50  ...   11.0     19.0          0   

        CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                    NaN         0           NaN          NaN      NaN   
1                    NaN         0           NaN          NaN      NaN   
2                    NaN         0           NaN          NaN      NaN   
3                    NaN         0           NaN          NaN      NaN   
4                    NaN         0           NaN          NaN      NaN   
...                  ...       ...           ...          ...      ...   
241100               NaN         0           3.0          0.0      0.0   
241101               NaN         0           0.0          0.0     82.0   
241102               NaN         0           3.0          0.0      0.0   
241103               NaN         0           NaN          NaN      NaN   
241104               NaN         0           0.0          0.0      1.0   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 NaN                NaN  
4                 NaN                NaN  
...               ...                ...  
241100            0.0               33.0  
241101            0.0                0.0  
241102            0.0               14.0  
241103            NaN                NaN  
241104            0.0               92.0  

[241105 rows x 29 columns]

We can identify each specific airplane using the "TailNum" column of this dataset. 

Develop a Python Pandas implementation that can find the most frequently used airplane within this dataset. 

Provide your Pandas code.
Provide the tail number of the most-used airplane.

In [15]:
flights.groupby(by = 'TailNum')['TailNum'].count().nlargest(5)

TailNum
N14937    935
N14939    934
N15948    885
N13935    842
N15941    818
Name: TailNum, dtype: int64

In [16]:
flights[['TailNum']].value_counts().nlargest(1)

TailNum
N14937     935
Name: count, dtype: int64

Over the entire dataset, which day of the week has the highest flight traffic (i.e. the busiest day of the week)?

Include canceled flights in your calculation. 

Provide your Pandas code.
Provide the busiest day of the week - is it Mo, Tu, Wed, Thu, Fri, Sat, or Sun?

In [17]:
flights.groupby(by = 'DayOfWeek')['DayOfWeek'].value_counts().nlargest(1)

DayOfWeek
5    36378
Name: count, dtype: int64

Is there a correlation between departure delay "DepDelay" and arrival delay "ArrDelay"?

Consider only flights that have a departure delay larger than one and an arrival delay larger than one.

In [18]:
large_delay = flights[(flights['DepDelay'] > 1) & (flights['ArrDelay'] > 1)]


In [19]:
large_delay['DepDelay'].corr(large_delay['ArrDelay'])

0.9324279653679205

a correlation coefficient of 0.932 indicates a very strong positive linear relationship between departure delay and arrival delay in the subset of flights where both departure delay and arrival delay are greater than 1.

This shows that the variables Arrival Delay and Departure Delay have a strong positive relationship, indicating that with an in increase in arrival delay, there is an increase in departure delay, and vice versa.
question 7

We will be considering the actual elapsed flight time "ActualElapsedTime" and the travel distance "Distance".

Create a simple linear regression model between these two values for predicting "ActualElapsedTime" using the "Distance" as the explanatory variable. 

Remove flights that do not have an "ActualElapsedTime" or "Distance" value, or their values are NaN or Infinity. 

Provide your Pandas code. 
What is the correlation coefficient between elapsed time and travel distance?
Provide your equation for this linear regression. 
Describe how well your model fits the data.
Provide the R-Squared of your model.

In [20]:
from sklearn.metrics import r2_score

cleaned = flights.dropna(subset=['Distance', 'ActualElapsedTime'])

X = cleaned[['Distance']]
y = cleaned['ActualElapsedTime']

model = LinearRegression()
model.fit(X, y)

print('Correlation Coefficient:', cleaned['Distance'].corr(cleaned['ActualElapsedTime']))

print('Equation: Actual Elapsed Time =', model.coef_[0], '* (Distance) + ', model.intercept_)

y_pred = model.predict(X)


print('R-Squared Score:', r2_score(y, y_pred))

Correlation Coefficient: 0.966758607067921
Equation: Actual Elapsed Time = 0.1275998228299878 * (Distance) +  30.76281009117305
R-Squared Score: 0.9346222043399015


In summary, an R-squared score of 0.935 indicates a very strong fit of the regression model to the data, suggesting that the model explains a large proportion of the variability in the dependent variable and has strong predictive power.






I want to fit a line to the sequence of points y = <101, 111, 122, 133, 144, 155> at the x-axis points x = <1, 2, 3, 4, 5, 6> respectively, using a gradient descent algorithm.

As in the class slides, the prediction is 
  , and we are using least-squares error for the loss.
The current estimate for the line is the intercept 
 , the slope 
 .
Our learning rate is 0.04.
After one additional iteration, what will be the estimate of the line? (i.e., what will be the new values of 
 and 
)?

In [21]:
# Given data
x = [1, 2, 3, 4, 5, 6]
y = [101, 111, 122, 133, 144, 155]

# Initial parameters
intercept = 50
slope = .2

# Learning rate
learning_rate = 0.04

# Number of data points
n = len(x)

# Perform one iteration of gradient descent
for i in range(n):
    # Compute predicted value for the ith data point
    y_pred_i = intercept + slope * x[i]
    
    # Compute the gradients of the loss function with respect to intercept and slope
    grad_intercept = -2 * (y[i] - y_pred_i)  # Partial derivative of loss w.r.t. intercept
    grad_slope = -2 * x[i] * (y[i] - y_pred_i)  # Partial derivative of loss w.r.t. slope
    
    # Update parameters using the gradients and learning rate
    intercept -= learning_rate * grad_intercept
    slope -= learning_rate * grad_slope

# Output the updated parameters
print("Updated intercept:", intercept)
print("Updated slope:", slope)


Updated intercept: 59.686041937510396
Updated slope: 16.621193213542405


In [30]:
X = np.array([1,2,3,4,5,6])
y = np.array([101, 111, 122, 133, 144, 155])

learning_rate = 0.04
max_itr = 1
precision = 0.01

m = .2
b = 50

n = len(X)

cost_list = []
itr_list = []


for i in range(max_itr):
    prediction = b + m * X
    cost = (1/n) * sum(( prediction - y)**2)
    m_gradient = (1.0/n) * np.sum(X * (prediction - y)) 
    b_gradient = (1.0/n) * np.sum(prediction - y)
    m = m - learning_rate * m_gradient
    b = b - learning_rate * b_gradient

    cost_list.append(cost)
    itr_list.append(i+1)
    print('Iteration:', i+1, '; Cost:', cost)

    if len(cost_list) > 1 and abs(cost_list[-1] - cost_list[-2]) <= precision:
        print("Stoped at iteration", i+1)
        break

    if len(itr_list) == max_itr:
        print("Stoped at iteration", i+1)

print('Coefficients:')
print('Slope:', m)
print('Intercept:', b)

Iteration: 1 ; Cost: 6255.206666666667
Stoped at iteration 1
Coefficients:
Slope: 12.218666666666666
Intercept: 53.07866666666666


1. Initialization:
   - I initialized the data:
     X = np.array([1, 2, 3, 4, 5, 6])
     y = np.array([101, 111, 122, 133, 144, 155])

   - I set hyperparameters:
     learning_rate = 0.04
     max_itr = 1
     precision = 0.01

   - I initialized the parameters for the linear regression model:
     m = 0.2
     b = 50



2. Gradient Descent Iteration:
   - I used one iteration of gradient descent:
     for i in range(max_itr):
   - Inside the loop, I calculated the predictions, cost (mean squared error), and gradients:
     prediction = b + m * X
     cost = (1/n) * sum((prediction - y)**2)
     m_gradient = (1.0/n) * np.sum(X * (prediction - y))
     b_gradient = (1.0/n) * np.sum(prediction - y)

   - I updated the parameters using the gradients and learning rate:
     m = m - learning_rate * m_gradient
     b = b - learning_rate * b_gradient



3. Output and Convergence:
   - During the iteration, I printed the iteration number and the cost:
     print('Iteration:', i+1, '; Cost:', cost)

   - I checked if the convergence criterion was met:
     if len(cost_list) > 1 and abs(cost_list[-1] - cost_list[-2]) <= precision:
         print("Stopped at iteration", i+1)
         break

   -The convergence criteria was met after the first iteration so the loop terminated early.

4. Solution:
   - I printed the final coefficients (slope and intercept) after the iteration:
     print('Coefficients:')
     print('Slope:', m)
     print('Intercept:', b)



5. Result:
   - After one iteration, I got the following coefficients:
     - Slope: 12.218666666666666
     - Intercept: 53.07866666666666
   - The cost after the first iteration was approximately 6255.206666666667.

My solution did one iteration of gradient descent and converged after the first iteration. The coefficients represent the linear regression model that best fits the given data after one iteration of gradient descent. 